File: iOS5bookExamples/ch36p901peopleGroups/p740peopleGroups/PeopleLister.m


#import "PeopleLister.h"
#import "Person.h"
#import "PeopleDocument.h"

@interface PeopleLister ()
@property (nonatomic, strong) NSMutableArray* people;
@property (nonatomic, strong) PeopleDocument* doc;
@property (nonatomic, strong) NSURL* fileURL;
@end

@implementation PeopleLister
@synthesize people, doc, fileURL=_fileURL;

- (id) initWithNibName:(NSString *)nibNameOrNil 
                bundle:(NSBundle *)nibBundleOrNil 
               fileURL: (NSURL*) fileURL {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self->_fileURL = fileURL;
        self.title = [fileURL.lastPathComponent stringByDeletingPathExtension];
        UIBarButtonItem* b = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(doAdd:)];
        self.navigationItem.rightBarButtonItems = [NSArray arrayWithObjects:b,nil];
    }
    return self;
}

- (id) initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    return nil; // we must not exist without a fileURL
}

- (void) viewDidLoad {
    [super viewDidLoad];
    [self.tableView registerNib:[UINib nibWithNibName:@"PersonCell" bundle:nil] forCellReuseIdentifier:@"Person"];
    
    NSFileManager* fm = [[NSFileManager alloc] init];
    self.doc = [[PeopleDocument alloc] initWithFileURL:self.fileURL];
    void (^listPeople) (BOOL) = ^(BOOL success) {
        if (success) {
            self.people = self.doc.people;
            [self.tableView reloadData];
        }
    };
    if (![fm fileExistsAtPath:[self.fileURL path]])
        [self.doc saveToURL:doc.fileURL 
           forSaveOperation:UIDocumentSaveForCreating 
          completionHandler:listPeople];
    else
        [self.doc openWithCompletionHandler:listPeople];
//    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(docChanged:) name:UIDocumentStateChangedNotification object:self.doc];
}

- (void) docChanged: (id) sender {
    // need to be a lot more sophisticated about this
    NSLog(@"state: %i", self.doc.documentState);
    if (self.doc.documentState == UIDocumentStateNormal) {
        [self.tableView endEditing:YES];
        self.people = self.doc.people;
        [self.tableView reloadData];
    }
}

- (void) doAdd: (id) sender {
    [self.tableView endEditing:YES];
    Person* newP = [Person new];
    [people addObject: newP];
    NSInteger ct = [people count];
    NSIndexPath* ix = [NSIndexPath indexPathForRow:ct-1 inSection:0];
    [self.tableView reloadData];
    [self.tableView scrollToRowAtIndexPath:ix atScrollPosition:UITableViewScrollPositionBottom animated:YES];
    UITableViewCell* cell = [self.tableView cellForRowAtIndexPath:ix];
    UITextField* tf = (UITextField*)[cell viewWithTag:1];
    [tf becomeFirstResponder];
    
    [self.doc updateChangeCount:UIDocumentChangeDone];
}

-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    if (!self.people)
        return 0; // no data yet
    return 1;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [people count];
}

-(UITableViewCell *)tableView:(UITableView *)tableView 
        cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = 
        [tableView dequeueReusableCellWithIdentifier:@"Person"];
    UITextField* first = (UITextField*)[cell viewWithTag:1];
    UITextField* last = (UITextField*)[cell viewWithTag:2];
    Person* p = [people objectAtIndex:indexPath.row];
    first.text = p.firstName;
    last.text = p.lastName;
    first.delegate = last.delegate = self;
    return cell;
}

-(void)textFieldDidEndEditing:(UITextField *)textField {
    NSLog(@"did end");
    UIView* v = textField.superview;
    while (![v isKindOfClass: [UITableViewCell class]])
        v = v.superview;
    UITableViewCell* cell = (UITableViewCell*)v;
    NSIndexPath* ip = [self.tableView indexPathForCell:cell];
    NSInteger row = ip.row;
    Person* p = [people objectAtIndex:row];
    [p setValue:textField.text forKey: ((textField.tag == 1) ? @"firstName" : @"lastName")];

    [self.doc updateChangeCount:UIDocumentChangeDone];
}

- (void)tableView:(UITableView *)tableView 
commitEditingStyle:(UITableViewCellEditingStyle)editingStyle 
forRowAtIndexPath:(NSIndexPath *)indexPath {
    [self.tableView endEditing:YES];
    [self.people removeObjectAtIndex:indexPath.row];
    [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] 
                     withRowAnimation:UITableViewRowAnimationAutomatic];

    [self.doc updateChangeCount:UIDocumentChangeDone];
}

- (void) forceSave: (id) n {
    NSLog(@"here");
    [self.tableView endEditing:YES];
    [self.doc saveToURL:doc.fileURL forSaveOperation:UIDocumentSaveForOverwriting completionHandler:nil];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(forceSave:) name:UIApplicationDidEnterBackgroundNotification object:nil];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self forceSave: nil];
    [[NSNotificationCenter defaultCenter] removeObserver: self];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver: self];
}


@end

File: iOS13bookExamples/bk2ch23p810DocumentBrowser/DocumentBrowser/PeopleLister.swift


import UIKit

class PeopleLister: UITableViewController, UITextFieldDelegate {
    
    let fileURL : URL
    var doc : PeopleDocument!
    var people : [Person] { // point to the document's model object
        get { return self.doc.people }
        set { self.doc.people = newValue }
    }

    init(fileURL:URL) {
        self.fileURL = fileURL
        super.init(nibName: "PeopleLister", bundle: nil)
    }
    
    required init(coder: NSCoder) {
        fatalError("NSCoding not supported")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = (self.fileURL.lastPathComponent as NSString)
            .deletingPathExtension
        let b = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(doAdd))
        self.navigationItem.rightBarButtonItems = [b]
        // we are now a presented view controller, so we also need a way to dismiss
        let b2 = UIBarButtonItem(title: "Done", style:.done, target: self, action: #selector(doDone))
        self.navigationItem.leftBarButtonItems = [b2]

        
        self.tableView.register(UINib(nibName: "PersonCell", bundle: nil), forCellReuseIdentifier: "Person")
        
        let fm = FileManager.default
        self.doc = PeopleDocument(fileURL:self.fileURL)
        
        func listPeople(_ success:Bool) {
            if success {
                // self.people = self.doc.people as NSArray as [Person]
                self.tableView.reloadData()
            }
        }
        if let _ = try? self.fileURL.checkResourceIsReachable() {
            self.doc.open(completionHandler:listPeople)
        } else {
            self.doc.save(to:self.doc.fileURL,
                          for: .forCreating,
                          completionHandler: listPeople)
        }
    }
    
    @objc func doAdd (_ sender: Any) {
        self.tableView.endEditing(true)
        let newP = Person(firstName: "", lastName: "")
        self.people.append(newP)
        let ct = self.people.count
        let ix = IndexPath(row:ct-1, section:0)
        self.tableView.reloadData()
        self.tableView.scrollToRow(at:ix, at:.bottom, animated:true)
        let cell = self.tableView.cellForRow(at:ix)!
        let tf = cell.viewWithTag(1) as! UITextField
        tf.becomeFirstResponder()
        
        self.doc.updateChangeCount(.done)
    }
    
    @objc func doDone (_ sender: Any) {
        self.presentingViewController?.dismiss(animated: true) {
            self.doc?.close(completionHandler: nil)
        }
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        if self.doc == nil {
            print("doc was nil")
            return 0
        }
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        print("self.people was \(self.people)")
        return self.people.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier:"Person", for: indexPath)
        let first = cell.viewWithTag(1) as! UITextField
        let last = cell.viewWithTag(2) as! UITextField
        let p = self.people[indexPath.row]
        first.text = p.firstName
        last.text = p.lastName
        first.delegate = self
        last.delegate = self
        return cell
    }
    
    func textFieldDidEndEditing(_ textField: UITextField) {
        print("did end editing")
        var v = textField.superview!
        while !(v is UITableViewCell) {v = v.superview!}
        let cell = v as! UITableViewCell
        let ip = self.tableView.indexPath(for:cell)!
        let row = ip.row
        let p = self.people[row]
        let which = textField.tag == 1 ? \Person.firstName : \.lastName
        p[keyPath:which] = textField.text!

        self.doc.updateChangeCount(.done)
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        self.tableView.endEditing(true)
        self.people.remove(at:indexPath.row)
        tableView.deleteRows(at:[indexPath], with:.automatic)
        
        self.doc.updateChangeCount(.done)
    }
    
    @objc func forceSave(_: Any?) {
        print("force save")
        self.tableView.endEditing(true)
        self.doc.save(to:self.doc.fileURL, for:.forOverwriting)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        NotificationCenter.default.addObserver(self, selector: #selector(forceSave), name: UIApplication.didEnterBackgroundNotification, object: nil)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.forceSave(nil)
        NotificationCenter.default.removeObserver(self)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }

}
